### PR TITLE
Implement merge endpoint

### PR DESCRIPTION
**Pull request by tomhoule**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,11 +5,15 @@
 
 ## Unreleased
 
+## [0.2.3] - 2017-08-31
+* Add a `/merge` endpoint that takes a MergeSpec and merges documents into a
+  single PDF file using `pdfunite` and ImageMagick.
```
Comment (peterfication) [inline CHANGELOG.md]: Add reason why

Comment (peterfication) [reply CHANGELOG.md]: in comparison to merge template

File: src/papers/merge.rs [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+use utils::s3::*;
+use utils::logging::file_logger;
+
+pub fn merge_documents(config: &'static Config, handle: &Handle, spec: MergeSpec) -> Box<Future<Item = (), Error = ()>> {
+    let pool = CpuPool::new(3);
+    let temp_dir = Temp::new_dir().expect("Could not create a temporary directory");
+    let max_asset_size = config.max_asset_size.clone();
+    let logger = file_logger(config.logger.clone(), temp_dir.as_ref());
+    let s3_prefix = s3_dir_name();
+    debug!(logger, "Downloading PDFs for mergin: {:?}", &spec.assets_urls);
```
Comment (peterfication) [inline src/papers/merge.rs]: merging

File: src/papers/merge.rs [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+use papers::MergeSpec;
+use http::*;
+use std::io::prelude::*;
+use slog::Logger;
+use std::path::*;
+use utils::callbacks::*;
+use utils::s3::*;
+use utils::logging::file_logger;
+
+pub fn merge_documents(config: &'static Config, handle: &Handle, spec: MergeSpec) -> Box<Future<Item = (), Error = ()>> {
```
Comment (peterfication) [inline src/papers/merge.rs]: a high level comment would be nice here

File: tests/merge/end_to_end.rs [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+    let mut core = tokio_core::reactor::Core::new().expect("could not create reactor");
+
+    let handle = core.handle();
+    let test_client = Client::new(&handle.clone());
+
+    let merge_spec = format!("{{
+        \"assets_urls\": [\"http://127.0.0.1:{port}/logo.png\", \"http://127.0.0.1:{port}/doc.pdf\"],
+        \"callback_url\": \"http://127.0.0.1:{port}/callback\",
+        \"variables\": {{
```
Comment (peterfication) [inline tests/merge/end_to_end.rs]: We don't need `variables` here, do we?

File: tests/toolbox/mod.rs [hunk 5]
```diff
@@ -0,0 +1,9 @@
+use rand::thread_rng;
+use rand::distributions::{Range, Sample};
+
+const PRIVATE_PORTS_MIN: u16 = 49152;
+const PRIVATE_PORTS_MAX: u16 = 65535;
+
+pub fn random_port() -> u16 {
```
Comment (peterfication) [inline tests/toolbox/mod.rs]: 👌

File: src/papers/merge.rs [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+            if output.status.success() {
+                debug!(logger, "{}", stdout);
+                Ok(output_path)
+            } else {
+                Err(ErrorKind::MergeFailed(stdout).into())
+            }
+        })
+    };
+
+    let presigned_url = {
```
Comment (peterfication) [inline src/papers/merge.rs]: Isn't it from here on the same as in the `submit` endpoint / in the `Renderer`?

File: tests/merge/end_to_end.rs (hunk 4 above)
Comment (tomhoule) [reply tests/merge/end_to_end.rs]: Indeed!

**Pull request merged**